src/mpid/common/thread/mpidu_thread_fallback.h

```diff
@@ -394,117 +306,61 @@ M*/
-        }                                                               \
-    } while (0)
-#define MPIDUI_THREAD_CS_YIELD_GLOBAL(mutex) do {} while (0)
-
-#elif MPICH_THREAD_GRANULARITY == MPICH_THREAD_GRANULARITY__VCI
-
-#define MPIDUI_THREAD_CS_YIELD_GLOBAL(mutex)                            \
-    do {                                                                \
-        if (MPIR_ThreadInfo.isThreaded) {                               \
-            int err_ = 0, equal_ = 0;                                   \
-            MPL_thread_id_t self_;                                      \
-            MPL_thread_self(&self_);                                    \
-            MPL_thread_same(&self_, &mutex.owner, &equal_);             \
-            if (equal_ && mutex.count > 0) {                            \
-                MPL_DBG_MSG(MPIR_DBG_THREAD, TYPICAL, "non-recursive yielding GLOBAL mutex"); \
-                MPL_DBG_MSG(MPIR_DBG_THREAD,VERBOSE,"enter MPIDU_Thread_yield"); \
-                MPIDU_Thread_yield(&mutex, &err_);                          \
-                MPL_DBG_MSG(MPIR_DBG_THREAD,VERBOSE,"exit MPIDU_Thread_yield"); \
-                MPIR_Assert(err_ == 0);                                 \
-            }                                                           \
-        }                                                               \
-    } while (0)
-#define MPIDUI_THREAD_CS_YIELD_VCI(mutex) MPIDUI_THREAD_CS_YIELD_GLOBAL(mutex)
-#define MPIDUI_THREAD_CS_YIELD_VCI_GLOBAL(mutex) do {} while (0)
-#define MPIDUI_THREAD_CS_YIELD_POBJ(mutex) do {} while (0)
+/* POBJ is only enabled with GRANULARITY__POBJ */
+#if MPICH_THREAD_GRANULARITY == MPICH_THREAD_GRANULARITY__POBJ
+#define MPIDUI_THREAD_CS_ENTER_POBJ(mutex)  MPIDUI_THREAD_CS_ENTER_NREC(mutex)
+#define MPIDUI_THREAD_CS_EXIT_POBJ(mutex)   MPIDUI_THREAD_CS_EXIT_NREC(mutex)
+#define MPIDUI_THREAD_CS_YIELD_POBJ(mutex)  MPIDUI_THREAD_CS_YIELD_NREC(mutex)
+#else
+#define MPIDUI_THREAD_CS_ENTER_POBJ(mutex)      /* NOOP */
+#define MPIDUI_THREAD_CS_EXIT_POBJ(mutex)       /* NOOP */
+#define MPIDUI_THREAD_CS_YIELD_POBJ(mutex)      /* NOOP */
+#endif
 
-#endif /* MPICH_THREAD_GRANULARITY */
+/* VCI is only enabled with MPICH_THREAD_GRANULARITY__VCI */
+#if MPICH_THREAD_GRANULARITY == MPICH_THREAD_GRANULARITY__VCI
+#define MPIDUI_THREAD_CS_ENTER_VCI(mutex) MPIDUI_THREAD_CS_ENTER_REC(mutex)
+#define MPIDUI_THREAD_CS_EXIT_VCI(mutex) MPIDUI_THREAD_CS_EXIT_REC(mutex)
+#define MPIDUI_THREAD_CS_YIELD_VCI(mutex) MPIDUI_THREAD_CS_YIELD_REC(mutex)
+#else
+#define MPIDUI_THREAD_CS_ENTER_VCI(mutex)       /* NOOP */
+#define MPIDUI_THREAD_CS_EXIT_VCI(mutex)        /* NOOP */
+#define MPIDUI_THREAD_CS_YIELD_VCI(mutex)       /* NOOP */
+#endif
 
-#else /* !defined(MPICH_IS_THREADED) */
+/* TRYENTER is of limited experimental use, only define the macros here when needed */
+#if MPICH_THREAD_GRANULARITY == MPICH_THREAD_GRANULARITY__VCI
```

Comment:
Could we keep everything in one place? Sure there might be some duplication but It is cleaner and easier to follow.

Reply:
done.